### DISCUSSION TITLE
Should Moose check block-restriction in FEProblemBase::duplicateVariableCheck

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#28560] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear Moose-Community,

Many of my Moose models consit of several subdomains / blocks while not all Variables and/or AuxVariables live on all subdomains/blocks. When setting up my models, I have repeatedly had the problem that variables/aux variables are required by different kernels, but were restricted to different blocks (which is of course a mistake on my part and/or other problems e.g. #28308). 
On the other hand, Moose checks in FEProblemBase::duplicateVariableCheck whether the variable already exists and is of the same type and thus avoids creating it twice. Unfortunately, Moose does not check that the block restriction is identical. Moose often outputs an error in the further modelling if the block-restriction was not identical, but the error messages often only indirectly indicating the underlying problem.

It would have been good if Moose had pointed out the problem of the inconsistent block restriction of the variables/aux variables more directly: Therefore, should Moose check block-restriction in FEProblemBase::duplicateVariableCheck and warn/error?
If so, I would like to collect your thoughts on whether this check should always be carried out or only on request.

https://github.com/idaholab/moose/blob/1926a677783ba66c803fb305330368d0960ff2a7/framework/src/problems/FEProblemBase.C#L2843-L2844

https://github.com/idaholab/moose/blob/1926a677783ba66c803fb305330368d0960ff2a7/framework/src/problems/FEProblemBase.C#L2540-L2551

**Comment by GiudGiud** [comment, marked as answer]
Hello

If you want the message to be more verbose and give more information (such as the block restriction) then that would work for us. Already it says "different type" but it would help the user to say which types are involved

However we can't really allow the check to let two "identical" variables be added without some serious checking. There are many variable parameters, we would have to check them all, not just the block restriction.

Guillaume

**Reply by jmeier** [reply]
I'll try to prepare an PR.
